compiler/rustc_middle/src/hir/map/mod.rs

```diff
@@ -1281,19 +1306,47 @@ pub(super) fn hir_module_items(tcx: TyCtxt<'_>, module_id: LocalDefId) -> Module
         }
 
         fn visit_trait_item(&mut self, item: &'hir TraitItem<'hir>) {
+            if associated_body(Node::TraitItem(item)).is_some() {
+                self.body_owners.push(item.def_id);
+            }
+
             self.trait_items.push(item.trait_item_id());
             intravisit::walk_trait_item(self, item)
         }
 
         fn visit_impl_item(&mut self, item: &'hir ImplItem<'hir>) {
+            if associated_body(Node::ImplItem(item)).is_some() {
+                self.body_owners.push(item.def_id);
+            }
+
             self.impl_items.push(item.impl_item_id());
             intravisit::walk_impl_item(self, item)
         }
 
         fn visit_foreign_item(&mut self, item: &'hir ForeignItem<'hir>) {
+            if associated_body(Node::ForeignItem(item)).is_some() {
+                self.body_owners.push(item.def_id);
+            }
+
             self.foreign_items.push(item.foreign_item_id());
             intravisit::walk_foreign_item(self, item)
         }
+
+        fn visit_expr(&mut self, ex: &'hir Expr<'hir>) {
+            if matches!(ex.kind, ExprKind::Closure { .. })
+                && associated_body(Node::Expr(ex)).is_some()
+            {
+                self.body_owners.push(ex.hir_id.owner);
```

Comment:
We want to record the `LocalDefId` of the closure.  `hir_id.owner` is the one of the closure's surrounding item-like (that we have already recorded in `visit_*item`).

```suggestion
                self.body_owners.push(self.tcx.hir().local_def_id(ex.hir_id));
```
Likewise for anon consts.